### ISSUE TITLE
zlib is not statically linked when building opencv_world

### ISSUE THREAD

**Issue by peters** [opencv/opencv#8355]
##### System information (version)
- OpenCV => [MASTER -  da0188a](https://github.com/opencv/opencv/tree/da0188acc0f7f14507a2db029e60779b324124f1)
- Operating System / Platform => Windows 10 64 Bit
- Compiler => Visual Studio 2017 (VC141)
- Cmake version => 3.8.0-RC2

##### Detailed description

When building `opencv_world` using Visual Studio 2017 `zlib` is not statically linked. @alalek you've made significant changes lately to the build process. Was this change intentional? 

![image](https://cloud.githubusercontent.com/assets/869/23821631/efebc7ee-0638-11e7-96f2-fe90be512055.png)

**Comment by peters** [comment]
Hmm.. Targeting Visual Studio 2015 (VC140) `zlib` is linked correctly:

![image](https://cloud.githubusercontent.com/assets/869/23805692/35e39300-05bf-11e7-9d71-d033fe841801.png)

**Comment by alalek** [comment]
> zlib is not statically linked

Builtin zlib adds as a [static library](https://github.com/opencv/opencv/blob/3.2.0/3rdparty/zlib/CMakeLists.txt#L76). No .dll should be created in this case.

If you using external zlib, then please check that ZLIB_LIBRARY/ZLIB_LIBRARIES CMake variable points the proper ".lib" file (instead of .dll implib stubs). List of variables are available in CMakeCache.txt / CMakeVars.txt files.

**Comment by peters** [comment]
That's the thing, I'm using zlib bundled in the 3rdparty folder. As per screenshots shown in the comments above. It only happens when using VS15 cmake generator. This is easily reproducible; Install VS2017 with c++ workload and then inspect the dll using Dependency Walker. With the new installer it should only you take around 20 minutes to verify. If you don't have time right now, do you have any pointers on where to start investigating? I looked at the Cmake issue tracker and there are multiple open issues related to VS2017 but they appear unrelated.

**Comment by peters** [comment]
Btw, zlib is being built as a library and I've verified that lib file present in the library output folder. It's only libpng and core that require zlib right? I downloaded zlib 1.2.8 from their girhub repository and manually built the library using VS2017. Then I copied the debug dll (zlibd.dll) into the same folder as opencv_world. Opencv then loads correctly.

**Comment by peters** [comment]
@alalek 

> If you using external zlib, then please check that ZLIB_LIBRARY/ZLIB_LIBRARIES CMake variable points the proper ".lib" file (instead of .dll implib stubs).

```
ZLIB_H=#define ZLIB_VERSION "1.2.8"
ZLIB_INCLUDE_DIRS=C:/Users/peters/Documents/GitHub/youpark.native/thirdparty/opencv/3rdparty/zlib;C:/Users/peters/Documents/GitHub/youpark.native/thirdparty/opencv/build/x64/Debug/3rdparty/zlib
ZLIB_LIBRARIES=zlib
ZLIB_LIBRARY=zlib
ZLIB_VERSION_MAJOR=1
ZLIB_VERSION_MINOR=2
ZLIB_VERSION_PATCH=8
ZLIB_VERSION_STRING=1.2.8
ZLIB_VERSION_TWEAK=
```

zlib is listed as a link target for `opencv_world`:
```
opencv_world_LIB_DEPENDS=general;ippicv;general;zlib;general;C:/Users/peters/Documents/GitHub/youpark.native/thirdparty/libjpeg-turbo/x64/jpeg.lib;general;libwebp;general;libpng;general;libtiff;general;libjasper;general;IlmImf;general;vfw32;general;comctl32;general;gdi32;general;ole32;general;setupapi;general;ws2_32;general;zlib;
```

**Comment by alalek** [comment]
Could you dump linker command for opencv_world320d.dll?

**Comment by alalek** [comment]
> It's only libpng and core that require zlib right

Could you please dump similar variables from CMakeCache.txt for "PNG"? (including BUILD_PNG)

**Comment by peters** [comment]
@alalek 

`opencv_world`:

`/OUT:"C:\Users\peters\Documents\GitHub\youpark.native\thirdparty\opencv\build\x64\Debug\bin\Debug\opencv_world320d.dll" /MANIFEST /NXCOMPAT /PDB:"C:/Users/peters/Documents/GitHub/youpark.native/thirdparty/opencv/build/x64/Debug/bin/Debug/opencv_world320d.pdb" /DYNAMICBASE "kernel32.lib" "user32.lib" "gdi32.lib" "winspool.lib" "shell32.lib" "ole32.lib" "oleaut32.lib" "uuid.lib" "comdlg32.lib" "advapi32.lib" "..\..\3rdparty\ippicv\ippicv_win\lib\intel64\ippicvmt.lib" "..\..\3rdparty\lib\Debug\zlibd.lib" "C:\Users\peters\Documents\GitHub\youpark.native\thirdparty\libjpeg-turbo\x64\jpeg.lib" "..\..\3rdparty\lib\Debug\libwebpd.lib" "..\..\3rdparty\lib\Debug\libpngd.lib" "..\..\3rdparty\lib\Debug\libtiffd.lib" "..\..\3rdparty\lib\Debug\libjasperd.lib" "..\..\3rdparty\lib\Debug\IlmImfd.lib" "vfw32.lib" "comctl32.lib" "setupapi.lib" "ws2_32.lib" /IMPLIB:"C:/Users/peters/Documents/GitHub/youpark.native/thirdparty/opencv/build/x64/Debug/lib/Debug/opencv_world320d.lib" /DLL /MACHINE:X64 /NODEFAULTLIB:"atlthunk.lib" /NODEFAULTLIB:"atlsd.lib" /NODEFAULTLIB:"libcmt.lib" /INCREMENTAL /PGD:"C:\Users\peters\Documents\GitHub\youpark.native\thirdparty\opencv\build\x64\Debug\bin\Debug\opencv_world320d.pgd" /SUBSYSTEM:CONSOLE /MANIFESTUAC:"level='asInvoker' uiAccess='false'" /ManifestFile:"opencv_world.dir\Debug\opencv_world320d.dll.intermediate.manifest" /ERRORREPORT:PROMPT /NOLOGO /LIBPATH:"C:/Users/peters/Documents/GitHub/youpark.native/thirdparty/opencv/3rdparty/lib" /LIBPATH:"C:/Users/peters/Documents/GitHub/youpark.native/thirdparty/opencv/3rdparty/lib/Debug" /TLBID:1 `

`zlib configuration`

![image](https://cloud.githubusercontent.com/assets/869/23822874/06e005f6-0656-11e7-8653-cf4d6ed9be33.png)

![image](https://cloud.githubusercontent.com/assets/869/23822915/13fe6dda-0657-11e7-9b08-5e49ebc545e2.png)

**Comment by peters** [comment]
`Variables related to PNG`:

`
BUILD_PNG=ON
HAVE_PNG=YES
PNG_DEFINITIONS=
PNG_INCLUDE_DIR=C:/Users/peters/Documents/GitHub/youpark.native/thirdparty/opencv/3rdparty/libpng
PNG_LIBPNG_VER_MAJOR=1
PNG_LIBPNG_VER_MINOR=6
PNG_LIBPNG_VER_RELEASE=24
PNG_LIBRARIES=libpng
PNG_LIBRARY=libpng
PNG_VERSION=1.6.24
PNG_VERSION_LINES=#define PNG_LIBPNG_VER_MAJOR   1;#define PNG_LIBPNG_VER_MINOR   6;#define PNG_LIBPNG_VER_RELEASE 24
libpng_BINARY_DIR=C:/Users/peters/Documents/GitHub/youpark.native/thirdparty/opencv/build/x64/Debug/3rdparty/libpng
libpng_LIB_DEPENDS=general;zlib;
libpng_SOURCE_DIR=C:/Users/peters/Documents/GitHub/youpark.native/thirdparty/opencv/3rdparty/libpng
`

`
/OUT:"C:\Users\peters\Documents\GitHub\youpark.native\thirdparty\opencv\build\x64\Debug\3rdparty\lib\Debug\libpngd.lib" /MACHINE:X64 /NOLOGO 
`

**Comment by alalek** [comment]
Thanks! There is external jpeg library - could you try to build without JPEG (-DWITH_JPEG=OFF)?

Could you upload somewhere opencv_world320d.dll and related .pdb file? I will try to find code/module which uses symbols from zlibd.dll

**Comment by peters** [comment]
PDB / OPENCV_WORLD debug build using builtin jpeg library.

[Debug.z01.zip](https://github.com/opencv/opencv/files/836492/Debug.z01.zip)
[Debug.zip](https://github.com/opencv/opencv/files/836491/Debug.zip)

Btw! You may have to rename `Debug.z01.zip` to `Debug.z01` in order to extract both files.

**Comment by alalek** [comment]
Unfortunatelly, I can't open this archive (Debug.zip doesn't have "PK" zlib prefix), Debug.z01 contains opencv_**ffmpeg**320_64.dll. Could you put opencv_**world** instead?

BTW, it is probably better to create commit in some temporary branch of your fork with mentioned binaries (current GitHub limit is 100Mb per commit, but you can create several commits).

**Comment by peters** [comment]
@alalek Sorry about that. Try this link:
https://youparkweb.blob.core.windows.net/peters/Debug.7z

**Comment by alalek** [comment]
Thank you for update!
Functions from zlibd.dll used by code from OpenCV (cvOpenFileStorage - gzopen) and libpng (called from functions with " using zstream" strings). So it is not caused by external dependency.
Could you upload zlibd `.lib`, `.dll` and `.pdb` files? Please dump zlib linker command from command line log too.

**Comment by peters** [comment]
@alalek Should a dll even exist since zlib is built as a static library? I cannot find a dll inside the `opencv` build folder. Please see attached zip file (.lib / .pdb)

**Comment by peters** [comment]
[Debug.zip](https://github.com/opencv/opencv/files/836637/Debug.zip)

**Comment by alalek** [comment]
DLL for shared libraries only. It is good that dll is missing.
Also this .lib file is static library, not implib for .dll. So it is very strange from where zlibd.dll is coming.

Could you try to rerun opencv_world linker command without `/INCREMENTAL` option (you may edit project properties in IDE or run linker command manually via coping this command into some ".cmd" file and running it)?

**Comment by peters** [comment]
@alalek Please see diagnostic build output.

[logfiles.zip](https://github.com/opencv/opencv/files/836667/logfiles.zip)

**Comment by peters** [comment]
@alalek Please see diagnostic build output with incremental linking turned off

[logfiles - incremental linking - off.zip](https://github.com/opencv/opencv/files/836804/logfiles.-.incremental.linking.-.off.zip)

**Comment by alalek** [comment]
I have installed MSVS2017 with CMake 3.8.0-rc2 and I'm not able to reproduce this issue,

CMake configuration:
```
-- General configuration for OpenCV 3.2.0-dev =====================================
--   Version control:               3.2.0-313-g502aa1f05
--
--   Platform:
--     Timestamp:                   2017-03-13T15:14:41Z
--     Host:                        Windows 10.0.10586 AMD64
--     CMake:                       3.8.0-rc2
--     CMake generator:             Visual Studio 15 2017 Win64
--     CMake build tool:            C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional/MSBuild/15.0/Bin/MSBuild.exe
--     MSVC:                        1910
--
--   C/C++:
--     Built as dynamic libs?:      YES
--     C++ Compiler:                C:/Program Files (x86)/Microsoft Visual Studio/2017/Professional/VC/Tools/MSVC/14.10.25017/bin/HostX86/x64/cl.exe  (ver 19.10.25017.0)
--     C++ flags (Release):         /DWIN32 /D_WINDOWS /W4 /GR /EHa  /D _CRT_SECURE_NO_DEPRECATE /D _CRT_NONSTDC_NO_DEPRECATE /D _SCL_SECURE_NO_WARNINGS /Gy /bigobj /Oi  /wd4251 /wd4324 /wd4275 /wd4589 /MP4  /MD /O2 /Ob2 /DNDEBUG  /Zi
--     C++ flags (Debug):           /DWIN32 /D_WINDOWS /W4 /GR /EHa  /D _CRT_SECURE_NO_DEPRECATE /D _CRT_NONSTDC_NO_DEPRECATE /D _SCL_SECURE_NO_WARNINGS /Gy /bigobj /Oi  /wd4251 /wd4324 /wd4275 /wd4589 /MP4  /MDd /Zi /Ob0 /Od /RTC1
--     C Compiler:                  C:/Program Files (x86)/Microsoft Visual Studio/2017/Professional/VC/Tools/MSVC/14.10.25017/bin/HostX86/x64/cl.exe
--     C flags (Release):           /DWIN32 /D_WINDOWS /W3  /D _CRT_SECURE_NO_DEPRECATE /D _CRT_NONSTDC_NO_DEPRECATE /D _SCL_SECURE_NO_WARNINGS /Gy /bigobj /Oi    /MP4  /MD /O2 /Ob2 /DNDEBUG  /Zi
--     C flags (Debug):             /DWIN32 /D_WINDOWS /W3  /D _CRT_SECURE_NO_DEPRECATE /D _CRT_NONSTDC_NO_DEPRECATE /D _SCL_SECURE_NO_WARNINGS /Gy /bigobj /Oi    /MP4  /MDd /Zi /Ob0 /Od /RTC1
--     Linker flags (Release):      /machine:x64  /INCREMENTAL:NO  /debug
--     Linker flags (Debug):        /machine:x64  /debug /INCREMENTAL
--     ccache:                      NO
--     Precompiled headers:         NO
--     Extra dependencies:          vfw32 comctl32 gdi32 ole32 setupapi ws2_32 C:/Program Files (x86)/IntelSWTools/compilers_and_libraries/windows/mkl/lib/intel64/mkl_intel_lp64.lib C:/Program Files (x86)/IntelSWTools/compilers_and_libraries/windows/mkl/lib/intel64/mkl_sequential.lib C:/Program Files (x86)/IntelSWTools/compilers_and_libraries/windows/mkl/lib/intel64/mkl_core.lib
--     3rdparty dependencies:       libjpeg libwebp libpng libtiff libjasper IlmImf zlib
--
--   OpenCV modules:
--     To be built:                 core flann imgproc ml photo video imgcodecs shape videoio highgui objdetect superres features2d calib3d stitching videostab world ts
--     Disabled:                    -
--     Disabled by dependency:      -
--     Unavailable:                 cudaarithm cudabgsegm cudacodec cudafeatures2d cudafilters cudaimgproc cudalegacy cudaobjdetect cudaoptflow cudastereo cudawarping cudev java python2 python3 viz
--
--   Windows RT support:            NO
--
--   GUI:
--     QT:                          NO
--     Win32 UI:                    YES
--     OpenGL support:              NO
--     VTK support:                 NO
--
--   Media I/O:
--     ZLib:                        build (ver 1.2.8)
--     JPEG:                        build (ver 90)
--     WEBP:                        build (ver encoder: 0x020e)
--     PNG:                         build (ver 1.6.24)
--     TIFF:                        build (ver 42 - 4.0.2)
--     JPEG 2000:                   build (ver 1.900.1)
--     OpenEXR:                     build (ver 1.7.1)
--     GDAL:                        NO
--     GDCM:                        NO
--
--   Video I/O:
--     Video for Windows:           YES
--     DC1394 1.x:                  NO
--     DC1394 2.x:                  NO
--     FFMPEG:                      YES (prebuilt binaries)
--       avcodec:                   YES (ver 57.64.101)
--       avformat:                  YES (ver 57.56.100)
--       avutil:                    YES (ver 55.34.100)
--       swscale:                   YES (ver 4.2.100)
--       avresample:                YES (ver 3.1.0)
--     GStreamer:                   NO
--     OpenNI:                      NO
--     OpenNI PrimeSensor Modules:  NO
--     OpenNI2:                     NO
--     PvAPI:                       NO
--     GigEVisionSDK:               NO
--     DirectShow:                  YES
--     Media Foundation:            NO
--     XIMEA:                       NO
--     Intel PerC:                  NO
--
--   Parallel framework:            Concurrency
--
--   Other third-party libraries:
--     Use IPP:                     9.0.1 [9.0.1]
--          at:                     C:/projects/build_vc15/3rdparty/ippicv/ippicv_win
--     Use IPP Async:               NO
--     Use Lapack:                  NO
--     Use Eigen:                   NO
--     Use Cuda:                    NO
--     Use OpenCL:                  YES
--     Use OpenVX:                  NO
--     Use custom HAL:              NO
--
--   OpenCL:                        <Dynamic loading of OpenCL library>
--     Include path:                C:/projects/opencv/3rdparty/include/opencl/1.2
--     Use AMDFFT:                  NO
--     Use AMDBLAS:                 NO
--
--   Python 2:
--     Interpreter:                 C:/Python27-64/python.exe (ver 2.7.12)
--
--   Python 3:
--     Interpreter:                 NO
--
--   Python (for build):            C:/Python27-64/python.exe
--
--   Java:
--     ant:                         NO
--     JNI:                         NO
--     Java wrappers:               NO
--     Java tests:                  NO
--
--   Matlab:                        Matlab not found or implicitly disabled
--
--   Documentation:
--     Doxygen:                     NO
--
--   Tests and samples:
--     Tests:                       YES
--     Performance tests:           YES
--     C/C++ Examples:              NO
--
--   Install path:                  C:/projects/build_vc15/install
--
--   cvconfig.h is in:              C:/projects/build_vc15
-- -----------------------------------------------------------------
```

Linker command:
```
Link:
  C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\VC\Tools\MSVC\14.10.25017\bin\HostX86\x64\link.exe /ERRORREPORT:QUEUE 
  /OUT:"C:\projects\build_vc15\bin\Debug\opencv_world320d.dll" /INCREMENTAL /NOLOGO /LIBPATH:C:/projects/opencv/3rdparty/lib /LIBPATH:C:/projects/opencv/3rdparty/lib/Debug
  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib ..\..\3rdparty\ippicv\ippicv_win\lib\intel64\ippicvmt.lib ..\..\3rdparty\lib\Debug\zlibd.lib ..\..\3rdparty\l
  ib\Debug\libjpegd.lib ..\..\3rdparty\lib\Debug\libwebpd.lib ..\..\3rdparty\lib\Debug\libpngd.lib ..\..\3rdparty\lib\Debug\libtiffd.lib ..\..\3rdparty\lib\Debug\libjasperd.lib ..
  \..\3rdparty\lib\Debug\IlmImfd.lib vfw32.lib comctl32.lib gdi32.lib ole32.lib setupapi.lib ws2_32.lib ..\..\3rdparty\lib\Debug\zlibd.lib "C:\Program Files (x86)\IntelSWTools\com
  pilers_and_libraries\windows\mkl\lib\intel64\mkl_intel_lp64.lib" "C:\Program Files (x86)\IntelSWTools\compilers_and_libraries\windows\mkl\lib\intel64\mkl_sequential.lib" "C:\Pro
  gram Files (x86)\IntelSWTools\compilers_and_libraries\windows\mkl\lib\intel64\mkl_core.lib" /NODEFAULTLIB:atlthunk.lib /NODEFAULTLIB:atlsd.lib /NODEFAULTLIB:libcmt.lib /MANIFEST
   /MANIFESTUAC:"level='asInvoker' uiAccess='false'" /manifest:embed /DEBUG /PDB:"C:/projects/build_vc15/bin/Debug/opencv_world320d.pdb" /SUBSYSTEM:CONSOLE /TLBID:1 /DYNAMICBASE /
  NXCOMPAT /IMPLIB:"C:/projects/build_vc15/lib/Debug/opencv_world320d.lib" /MACHINE:X64   /machine:x64 /DLL opencv_world.dir\Debug\algorithm.obj
  opencv_world.dir\Debug\alloc.obj
  opencv_world.dir\Debug\arithm.obj
... more objs ...
  opencv_world.dir\Debug\/__/videostab/src/optical_flow.cpp.obj
     Creating library C:/projects/build_vc15/lib/Debug/opencv_world320d.lib and object C:/projects/build_vc15/lib/Debug/opencv_world320d.exp
  opencv_world.vcxproj -> C:\projects\build_vc15\bin\Debug\opencv_world320d.dll
```

Dependency walker:
![opencv_world_vs2017](https://cloud.githubusercontent.com/assets/4981219/23861709/ea3188cc-081b-11e7-8c62-8489d1fb7929.png)

**Comment by peters** [comment]
@alalek That's unfortunate. Leave this issue open and I'll try to get back to you in a couple of days :)

**Comment by kikaxa** [comment]
I had the same issue, it disappeared after recreating from clean configuration and updating to msvc17.2

**Comment by peters** [comment]
@kikaxa Is that a preview release?

**Comment by kikaxa** [comment]
@peters  opencv 3.2, cmake 3.7, qt creator 4.3, msvc2017.2

**Comment by peters** [comment]
@alalek The problem disappeared after updating to Windows 10 to `Creators Update`.

**Comment by peters** [comment]
@alalek After upgrading to latest master this issue yet again resurfaced. Tried upgrading cmake to `3.12.0`  and it throws the following error: `Invalid escape sequence \u`. I reviewed our build environment and of course there was an invalid character sequence present in the path of `libjpeg-turbo` (not using opencv bundled version). Removing the erroneous character and cmake configuration succeeds. I guess `zlib` was randomly affected because it it's listed as the first `ocv` module to link.
